tensorflow/core/kernels/matrix_triangular_solve_op.cc

```diff
@@ -76,40 +174,89 @@ class MatrixTriangularSolveOp : public LinearAlgebraOp<Scalar> {
       // an empty set of equation as the empty matrix.
       return;
     }
-    const Scalar min_abs_pivot = matrix.diagonal().cwiseAbs().minCoeff();
-    OP_REQUIRES(context, min_abs_pivot > Scalar(0),
-                errors::InvalidArgument("Input matrix is not invertible."));
+
+    auto matrix_ptr = AsDeviceMemory(matrix.data());
+    auto rhs_ptr = AsDeviceMemory(rhs.data());
+    auto out_ptr = AsDeviceMemory(output.data());
+
+    auto* stream = context->op_device_context()->stream();
+    uint64 belems = rhs.rows() * rhs.cols();
+    bool copy_status =
+        stream->ThenMemcpyD2D(&out_ptr, rhs_ptr, sizeof(Scalar) * belems).ok();
+    if (!copy_status) {
+      context->SetStatus(
+          errors::Internal("Failed to copy B into output before TRSM"));
+    }
+    perftools::gputools::blas::UpperLower uplo;
+    perftools::gputools::blas::Transpose trans;
     if (lower_) {
-      auto triangle = matrix.template triangularView<Eigen::Lower>();
-      if (adjoint_) {
-        output.noalias() = triangle.adjoint().solve(rhs);
-      } else {
-        output.noalias() = triangle.solve(rhs);
-      }
+      uplo = perftools::gputools::blas::UpperLower::kUpper;
+    } else {
+      uplo = perftools::gputools::blas::UpperLower::kLower;
+    }
+    if (adjoint_) {
+      trans = perftools::gputools::blas::Transpose::kTranspose;
     } else {
-      auto triangle = matrix.template triangularView<Eigen::Upper>();
-      if (adjoint_) {
-        output.noalias() = triangle.adjoint().solve(rhs);
-      } else {
-        output.noalias() = triangle.solve(rhs);
-      }
+      trans = perftools::gputools::blas::Transpose::kNoTranspose;
```

Comment:
can you use longer variable names instead of 'trans', 'lda', 'ldb', etc?

then in ThenBlasTrsm call, you can do e.g. upper_or_lower /* uplo */, ..., etc.

easier for other users to read.